### PR TITLE
Add argument postGenerationCommand to ExternalLHEProducer and provide mergeLHE.py script

### LINKED ISSUE #33544: Segfault in Herwig
Workflow 535.0 step 1 crashes in CMSSW_12_0_X_2021-04-27-1100
```
27-Apr-2021 14:34:58 CEST  Successfully opened LHE file thread0/cmsgrid_final.lhe
Error: No such file or directory: cmsgrid_final.lhe
Error: The object '/Herwig/Partons/PDFSet_nnlo' was not created as another object with that name already exists.
Error: The object '/Herwig/Partons/PDFSet_lo' was not created as another object with that name already exists.
Error: The object '/Herwig/EventHandlers/LesHouchesHandler' was not created as another object with that name already exists.
Error: The object '/Herwig/Cuts/NoCuts' was not created as another object with that name already exists.
Error: The object '/Herwig/Partons/LHAPDF' was not created as another object with that name already exists.
Error: The object '/Herwig/EventHandlers/LesHouchesReader' was not created as another object with that name already exists.
Error: No such file or directory: cmsgrid_final.lhe
Herwig: EventGenerator not available.
Check if 'InterfaceMatchboxTest.run' is a valid run file.


A fatal system signal has occurred: segmentation violation
The following is the call stack containing the origin of the signal.

Tue Apr 27 14:35:01 CEST 2021
Thread 5 (Thread 0x2b989fe00700 (LWP 20760)):
#2  0x00002b9869c53720 in sig_pause_for_stacktrace () from /cvmfs/cms-ib.cern.ch/nweek-02678/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_X_2021-04-25-0000/lib/slc7_amd64_gcc900/pluginFWCoreServicesPlugins.so
#3  <signal handler called>
#4  0x00002b985ec41d19 in syscall () from /lib64/libc.so.6
#5  0x00002b985db1ed9d in tbb::internal::futex_wait (comparand=2, futex=0x2b9865f0302c) at ../../include/tbb/machine/linux_common.h:81

Thread 4 (Thread 0x2b989ef70700 (LWP 20759)):
#2  0x00002b9869c53720 in sig_pause_for_stacktrace () from /cvmfs/cms-ib.cern.ch/nweek-02678/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_X_2021-04-25-0000/lib/slc7_amd64_gcc900/pluginFWCoreServicesPlugins.so
#3  <signal handler called>
#4  0x00002b985ec41d19 in syscall () from /lib64/libc.so.6
#5  0x00002b985db1ed9d in tbb::internal::futex_wait (comparand=2, futex=0x2b9865f0312c) at ../../include/tbb/machine/linux_common.h:81

Thread 3 (Thread 0x2b989e56f700 (LWP 20758)):
#2  0x00002b9869c53720 in sig_pause_for_stacktrace () from /cvmfs/cms-ib.cern.ch/nweek-02678/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_X_2021-04-25-0000/lib/slc7_amd64_gcc900/pluginFWCoreServicesPlugins.so
#3  <signal handler called>
#4  0x00002b985ec41d19 in syscall () from /lib64/libc.so.6
#5  0x00002b985db1ed9d in tbb::internal::futex_wait (comparand=2, futex=0x2b9865f030ac) at ../../include/tbb/machine/linux_common.h:81

Thread 1 (Thread 0x2b9860844180 (LWP 20111)):
#2  0x00002b9869c5456c in edm::service::InitRootHandlers::stacktraceFromThread() () from /cvmfs/cms-ib.cern.ch/nweek-02678/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_X_2021-04-25-0000/lib/slc7_amd64_gcc900/pluginFWCoreServicesPlugins.so
#3  0x00002b9869c55922 in sig_dostack_then_abort () from /cvmfs/cms-ib.cern.ch/nweek-02678/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_X_2021-04-25-0000/lib/slc7_amd64_gcc900/pluginFWCoreServicesPlugins.so
#4  <signal handler called>
#5  0x00002b98909476d0 in (anonymous namespace)::HerwigGenericRun(Herwig::HerwigUI const&, bool) () from /cvmfs/cms-ib.cern.ch/nweek-02678/slc7_amd64_gcc900/cms/cmssw-patch/CMSSW_12_0_X_2021-04-27-1100/external/slc7_amd64_gcc900/lib/libHerwigAPI.so.2
#6  0x00002b9890949757 in Herwig::API::prepareRun(Herwig::HerwigUI const&) () from /cvmfs/cms-ib.cern.ch/nweek-02678/slc7_amd64_gcc900/cms/cmssw-patch/CMSSW_12_0_X_2021-04-27-1100/external/slc7_amd64_gcc900/lib/libHerwigAPI.so.2
#7  0x00002b9890379abd in Herwig7Interface::callHerwigGenerator() () from /cvmfs/cms-ib.cern.ch/nweek-02678/slc7_amd64_gcc900/cms/cmssw-patch/CMSSW_12_0_X_2021-04-27-1100/lib/slc7_amd64_gcc900/libGeneratorInterfaceHerwig7Interface.so
#8  0x00002b9890379cb1 in Herwig7Interface::initGenerator() () from /cvmfs/cms-ib.cern.ch/nweek-02678/slc7_amd64_gcc900/cms/cmssw-patch/CMSSW_12_0_X_2021-04-27-1100/lib/slc7_amd64_gcc900/libGeneratorInterfaceHerwig7Interface.so
#9  0x00002b98903390f4 in Herwig7Hadronizer::initializeForInternalPartons() () from /cvmfs/cms-ib.cern.ch/nweek-02678/slc7_amd64_gcc900/cms/cmssw-patch/CMSSW_12_0_X_2021-04-27-1100/lib/slc7_amd64_gcc900/pluginGeneratorInterfaceHerwig7HadronizerPlugins.so
#10 0x00002b9890346444 in edm::GeneratorFilter<Herwig7Hadronizer, gen::ExternalDecayDriver>::beginLuminosityBlockProduce(edm::LuminosityBlock&, edm::EventSetup const&) () from /cvmfs/cms-ib.cern.ch/nweek-02678/slc7_amd64_gcc900/cms/cmssw-patch/CMSSW_12_0_X_2021-04-27-1100/lib/slc7_amd64_gcc900/pluginGeneratorInterfaceHerwig7HadronizerPlugins.so
#11 0x00002b985c488acd in edm::one::EDFilterBase::doBeginLuminosityBlock(edm::LumiTransitionInfo const&, edm::ModuleCallingContext const*) () from /cvmfs/cms-ib.cern.ch/nweek-02678/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_X_2021-04-25-0000/lib/slc7_amd64_gcc900/libFWCoreFramework.so
#12 0x00002b985c466d10 in edm::WorkerT<edm::one::EDFilterBase>::implDoBegin(edm::LumiTransitionInfo const&, edm::ModuleCallingContext const*) () from /cvmfs/cms-ib.cern.ch/nweek-02678/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_X_2021-04-25-0000/lib/slc7_amd64_gcc900/libFWCoreFramework.so
#13 0x00002b985c370e96 in decltype ({parm#1}()) edm::convertException::wrap<edm::Worker::runModule<edm::OccurrenceTraits<edm::LuminosityBlockPrincipal, (edm::BranchActionType)0> >(edm::OccurrenceTraits<edm::LuminosityBlockPrincipal, (edm::BranchActionType)0>::TransitionInfoType const&, edm::StreamID, edm::ParentContext const&, edm::OccurrenceTraits<edm::LuminosityBlockPrincipal, (edm::BranchActionType)0>::Context const*)::{lambda()#1}>(edm::Worker::runModule<edm::OccurrenceTraits<edm::LuminosityBlockPrincipal, (edm::BranchActionType)0> >(edm::OccurrenceTraits<edm::LuminosityBlockPrincipal, (edm::BranchActionType)0>::TransitionInfoType const&, edm::StreamID, edm::ParentContext const&, edm::OccurrenceTraits<edm::LuminosityBlockPrincipal, (edm::BranchActionType)0>::Context const*)::{lambda()#1}) () from /cvmfs/cms-ib.cern.ch/nweek-02678/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_X_2021-04-25-0000/lib/slc7_amd64_gcc900/libFWCoreFramework.so
#14 0x00002b985c37109d in bool edm::Worker::runModule<edm::OccurrenceTraits<edm::LuminosityBlockPrincipal, (edm::BranchActionType)0> >(edm::OccurrenceTraits<edm::LuminosityBlockPrincipal, (edm::BranchActionType)0>::TransitionInfoType const&, edm::StreamID, edm::ParentContext const&, edm::OccurrenceTraits<edm::LuminosityBlockPrincipal, (edm::BranchActionType)0>::Context const*) () from /cvmfs/cms-ib.cern.ch/nweek-02678/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_X_2021-04-25-0000/lib/slc7_amd64_gcc900/libFWCoreFramework.so
#15 0x00002b985c371346 in std::__exception_ptr::exception_ptr edm::Worker::runModuleAfterAsyncPrefetch<edm::OccurrenceTraits<edm::LuminosityBlockPrincipal, (edm::BranchActionType)0> >(std::__exception_ptr::exception_ptr const*, edm::OccurrenceTraits<edm::LuminosityBlockPrincipal, (edm::BranchActionType)0>::TransitionInfoType const&, edm::StreamID, edm::ParentContext const&, edm::OccurrenceTraits<edm::LuminosityBlockPrincipal, (edm::BranchActionType)0>::Context const*) () from /cvmfs/cms-ib.cern.ch/nweek-02678/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_X_2021-04-25-0000/lib/slc7_amd64_gcc900/libFWCoreFramework.so
#16 0x00002b985c3718b0 in void edm::SerialTaskQueueChain::actionToRun<edm::Worker::RunModuleTask<edm::OccurrenceTraits<edm::LuminosityBlockPrincipal, (edm::BranchActionType)0> >::execute()::{lambda()#1}&>(edm::Worker::RunModuleTask<edm::OccurrenceTraits<edm::LuminosityBlockPrincipal, (edm::BranchActionType)0> >::execute()::{lambda()#1}&) () from /cvmfs/cms-ib.cern.ch/nweek-02678/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_X_2021-04-25-0000/lib/slc7_amd64_gcc900/libFWCoreFramework.so
#17 0x00002b985c371a61 in edm::SerialTaskQueue::QueuedTask<edm::SerialTaskQueueChain::push<edm::Worker::RunModuleTask<edm::OccurrenceTraits<edm::LuminosityBlockPrincipal, (edm::BranchActionType)0> >::execute()::{lambda()#1}&>(tbb::task_group&, edm::Worker::RunModuleTask<edm::OccurrenceTraits<edm::LuminosityBlockPrincipal, (edm::BranchActionType)0> >::execute()::{lambda()#1}&)::{lambda()#1}>::execute() () from /cvmfs/cms-ib.cern.ch/nweek-02678/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_X_2021-04-25-0000/lib/slc7_amd64_gcc900/libFWCoreFramework.so
#18 0x00002b985c107f49 in tbb::internal::function_task<edm::SerialTaskQueue::spawn(edm::SerialTaskQueue::TaskBase&)::{lambda()#1}>::execute() () from /cvmfs/cms-ib.cern.ch/nweek-02678/slc7_amd64_gcc900/cms/cmssw/CMSSW_12_0_X_2021-04-25-0000/lib/slc7_amd64_gcc900/libFWCoreConcurrency.so


Current Modules:
Module: Herwig7GeneratorFilter:generator (crashed)
Module: none
Module: none
Module: none
```
https://cmssdt.cern.ch/SDT/cgi-bin/logreader/slc7_amd64_gcc900/CMSSW_12_0_X_2021-04-27-1100/pyRelValMatrixLogs/run/535.0_TTbar_13TeV_Pow_herwig7+TTbar_13TeV_Pow_herwig7+HARVESTGEN/step1_TTbar_13TeV_Pow_herwig7+TTbar_13TeV_Pow_herwig7+HARVESTGEN.log#/

### PR DESCRIPTION
**Pull request by colizz**
#### PR description:

- Add an argument `postGenerationCommand` to ExternalLHEProducer module that can run a self-specified command right after LHE event generation (i.e. launched by the `ScriptName`).
- Add a `mergeLHE.py` script that merges LHEs using cmssw utility.
- Resolves: #33544 
  Now we provide `postGenerationCommand` to merge LHE events after concurrent LHE event generation for the later use of H7GeneratorFilter.

#### PR validation:

Pass workflow 535.0

### CODE REVIEW

**Comment by qliphy** [comment]
please test workflow 511.0, 535.0

**Comment by qliphy** [comment]
type bugfix

**Comment by qliphy** [comment]
"RelVals-INPUT   4000.0DAS Error"  not relevant to this PR.   There is no such issue in other recent PR tests.

**Comment by qliphy** [comment]
please test

**Comment by mrodozov** [comment]
@colizz if you change this
Solve #33544:
in your first comment (the so called PR description)
to
Resolves: #33544 
the issue will be closed when this is merged

**Comment by qliphy** [comment]
urgent

This is needed to resolve a crash in IBs  @cms-sw/generators-l2

**Comment by agrohsje** [comment]
@colizz did you check that the output from running the merge for 2 lhe files and a single file is the same? There is some info that depends on the number of events, like e.g. MG5 has this average and sum as event norm.

**Comment by agrohsje** [comment]
@qliphy do you prefer to merge now and disentangle physics from a technical fix?

**Comment by colizz** [comment]
> @colizz did you check that the output from running the merge for 2 lhe files and a single file is the same? There is some info that depends on the number of events, like e.g. MG5 has this average and sum as event norm.

Indeed I haven't yet validated its physics result. My previous thought was that since we are using cmssw utility to merge and write out an LHE file, it works just the same as cmssw's intrinsic mechanism to pass EDM LHE event one by one to the hadronizer.

However, it comes to me we are now fixing H7 issue, and H7 digests LHE directly, so there indeed may cause a difference... Even in the future we notice a issue, what we may do is to correct cmssw's way of handling LHE (I think should be in `LHERuninfo.cc`, `LHEEvent.cc`..) So I would de-correlate this PR with cmssw's way of handling LHE. What do you think?

BTW I will handle with the validation now. (And, if in future we move to H7HadrionzerFilter, we are actually facing the same validation task).

**Comment by agrohsje** [comment]
Hi @colizz. What do you mean with "de-correlate this PR with cmssw's way of handling LHE". Do you suggest to fix physics afterwards in a separate PR.

**Comment by colizz** [comment]
> Hi @colizz. What do you mean with "de-correlate this PR with cmssw's way of handling LHE". Do you suggest to fix physics afterwards in a separate PR.

Hi @agrohsje, yes that is what I mean. 
Of course, we may also keep the PR open until we finalise the physics validation. (I think it will take ~2 weeks).

**Comment by agrohsje** [comment]
If @qliphy is ok with merging and validating and fixing physics afterwards, I am fine too.

**Comment by qliphy** [comment]
thanks @colizz @agrohsje 
We may also keep this PR open and fix the IB issue for the moment by forcing externalLHEProducer.generateConcurrently = cms.untracked.bool(False), in TT_13TeV_Pow_Herwig7_cff.py
https://github.com/cms-sw/cmssw/issues/33544#issuecomment-828281661

**Comment by agrohsje** [comment]
Thanks @qliphy. Will sign #33628 as soon as tests are done.

**Comment by silviodonato** [comment]
@colizz do you have any news on this PR?
(#33628 has been merged meanwhile)

**Comment by colizz** [comment]
Hi @silviodonato , sorry but not yet. I will be able to handle it by the end of week

**Comment by colizz** [comment]
> Hi @silviodonato , sorry but not yet. I will be able to handle it by the end of week

four days more... I overestimate the tasks on the todo list. Hopes it won't affect the schedule.

**Comment by colizz** [comment]
Sorry for the long delay. I am now facing an issue that may need some suggestions from the GEN group.

I noticed that we are in an awkward situation that `LHEEventProduct` does not seem to contain all LHE information. For each event block from LHE, `LHEEventProduct` will store its `<scales>` and `<rwgt>` sub-block, but does not seem to store `<clustering>` (please see (*) as an example event block generated by MG5). But it seems to me that the latter will be utilised by Herwig7 in FxFxHandler module. If this is true, it means we cannot convert an `LHEEventProduct` to an LHE file used as Herwig's input.

I think we will probably need to use *external* scripts to merge LHE. For MG-produced LHE I understand we can use this [merge.pl](https://github.com/cms-sw/genproductions/blob/master/bin/MadGraph5_aMCatNLO/Utilities/merge.pl) script. But I am not sure which to use for merging Powheg-produced LHE. @agrohsje , may I ask for your recommendation on this? Besides, do we have other situations that an ME generator (other than MG5 or H7) is interfaced with Herwig7?

Thank you very much in advance.

(*)
```
<event>
 4      0 +2.5657158e+03 1.02456800e+02 7.54677100e-03 1.15961800e-01
       21 -1    0    0  503  501 +0.0000000000e+00 +0.0000000000e+00 +4.8073731891e+02 4.8073731891e+02 0.0000000000e+00 0.0000e+00 -1.0000e+00
       21 -1    0    0  504  502 -0.0000000000e+00 -0.0000000000e+00 -1.1744824940e+02 1.1744824940e+02 0.0000000000e+00 0.0000e+00 1.0000e+00
       21  1    1    2  503  502 +9.9010159199e+01 +2.6351184538e+01 -8.8215984461e+01 1.3520154017e+02 0.0000000000e+00 0.0000e+00 1.0000e+00
       21  1    1    2  504  501 -9.9010159199e+01 -2.6351184538e+01 +4.5150505397e+02 4.6298402814e+02 0.0000000000e+00 0.0000e+00 -1.0000e+00
<scales pt_clust_3="102.45680" pt_clust_4="13000.00000"></scales>
<mgrwt>
<rscale>  1 0.10245680E+03</rscale>
<asrwt>  1 0.10245680E+03</asrwt>
<pdfrwt beam="1">  1       21 0.73959586E-01 0.10245680E+03</pdfrwt>
<pdfrwt beam="2">  1       21 0.18068962E-01 0.10245680E+03</pdfrwt>
<totfact> 0.71019083E+02</totfact>
</mgrwt>
<clustering>
<clus scale="  102.457">  2  3  1 -1</clus>
<clus scale="  102.457">  1  3  2 -1</clus>
</clustering>
<rwgt>
<wgt id='1001'> +2.5657158e+03 </wgt>
<wgt id='1002'> +2.1930414e+03 </wgt>
...
</rwgt>
</event>
```

**Comment by colizz** [comment]
Hi, as a simple solution I reimplement the `mergeLHE.py` script. Now it runs separately with cmssw, and includes all <event> info without losing any tags. 

The default merging mode "copies the header of the first LHE file, merges and outputs the init block, then concatenates all event blocks". Besides, there is also an MG5-specific mode using the MG's merge.pl script to do merging.

The possible consequence to the resulted LHE is summarized [in this slide](https://coli.web.cern.ch/coli/tmp/.210523-233804_mergeLHE/mergeLHE_illustration.pdf) - overall it won't cause any difference to LHEEventProduct.

We also compare GEN-level plots and see agreement with single-core vs. concurrent mode.

File: GeneratorInterface/LHEInterface/scripts/mergeLHE.py [hunk 1]
```diff
@@ -2,10 +2,149 @@
...
+                    fw.write(_line)
+            fw.write('</LesHouchesEvents>\n')
+            os.remove('.tmp.lhe')
+
+
+class MadGraph5LHEMerger(BaseLHEMerger):
+    """Use the merger script in genproductions dedicated for MadGraph5-produced LHEs"""
+    def __init__(self, inputFiles, outputFileName):
+        super(MadGraph5LHEMerger, self).__init__(inputFiles, outputFileName)
+        self._merger_script_url = 'https://raw.githubusercontent.com/cms-sw/genproductions/5c1e865a6fbe3a762a28363835d9a804c9cf0dbe/bin/MadGraph5_aMCatNLO/Utilities/merge.pl'
```
Comment (qliphy) [inline GeneratorInterface/LHEInterface/scripts/mergeLHE.py:140]: Shouldn't it be https://raw.githubusercontent.com/cms-sw/genproductions/master/bin/MadGraph5_aMCatNLO/Utilities/merge.pl ?

Comment (colizz) [reply GeneratorInterface/LHEInterface/scripts/mergeLHE.py:140]: Yes this is a consideration of stability. We use a fixed version of the script.

**Comment by agrohsje** [comment]
Hi @colizz . Wouldn't -forceMadGraph5LHEMerger suffer from the problem we discussed yesterday? I.e. cannot handle different number of events in input files and sum as event norm? What about the python implementation. Did you run a test to see that 500 and 1000 events give identical weights as 1500 events in one go. It should for average, what about sum? 
Should we decouple this PR from the H7 problem of LHEEventInfoProduct and GenEventInfoProduct mismatch?

**Comment by colizz** [comment]
> Hi @colizz . Wouldn't -forceMadGraph5LHEMerger suffer from the problem we discussed yesterday? I.e. cannot handle different number of events in input files and sum as event norm? What about the python implementation. Did you run a test to see that 500 and 1000 events give identical weights as 1500 events in one go. It should for average, what about sum?

Hi @agrohsje , in the python implementation the event weight XWGTUP is not modified, only the <init> block is changed. In the case of combining 500 and 1000 uevents sing "sum" as event norm, the per-event XWGTUP indeed inherits from their old value and thus be different. In fact I don't think it may cause an impact on hadronization results (in principle, it should be the same with feeding the un-merged LHEs separately to the hadronizer). But as a goal of designing a reasonable merging script I think I'll need to handle this case. Will revise the script soon.

> Should we decouple this PR from the H7 problem of LHEEventInfoProduct and GenEventInfoProduct mismatch?

Yes I think so - this may need some fix intrinsic to H7GeneratorFilter module (and a naive idea from my side is that it's only possible in the HadroinzerFilter prototype).

**Comment by colizz** [comment]
Hello, sorry for the interval. I am able to centrally work on this PR. It seems I cannot revert #33628 in this PR (and given it has lasted for so long), so maybe I will close this and open a new one.

**Pull request closed without merging**